src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -3217,37 +3217,105 @@ private void CompileTrappableExpression(List<Expression> exprList, StatementAst
             var expr = Compile(stmt);
```

Comment:
@rjmholt, your last commit had 1 failures in `PowerShell-CI-macos`
Tests for setting $? for execution success.Sets $? for case 'Write-Error "Bad"; $(); Hook $?'
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Language/Scripting/DollarHook.Tests.ps1: line 68
68:         $script:hookValues | Should -Be $Results
```